evm/src/cpu/kernel/asm/bignum/modmul.asm

```diff
@@ -0,0 +1,168 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+// Stores a * b % m in output_loc, leaving a, b, and m unchanged.
+// a, b, and m must have the same length.
+// Both output_loc and scratch_1 must have size length.
+// All of scratch_2, scratch_3, and scratch_4 must have size 2 * length and be initialized with zeroes.
```

Comment:
I'm not sure whether this would be an improvement overall, but note that you can remove the need for additional prover inputs by using [Barrett reduction](https://en.wikipedia.org/wiki/Barrett_reduction) which uses a precomputed "pseudo-inverse" of the modulus. The pseudo-inverse can be easily computed (once at the start of the `MODEXP`) using [the trick described here](https://medium.com/wicketh/mathemagic-512-bit-division-in-solidity-afa55870a65).